classes/observers/listeners/SendSubmissionAcknowledgement.php

```diff
@@ -0,0 +1,126 @@
+<?php
+/**
+ * @file classes/observers/listeners/SendSubmissionAcknowledgement.php
+ *
+ * Copyright (c) 2014-2022 Simon Fraser University
+ * Copyright (c) 2000-2022 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class SendSubmissionAcknowledgement
+ * @ingroup observers_listeners
+ *
+ * @brief Send an email acknowledgement to the submitting author when a new submission is submitted
+ *
+ * Sends an email to all users with author stage assignments and
+ * sends a separate email to all other contributors named on the
+ * submission.
+ */
+
+namespace PKP\observers\listeners;
+
+use APP\author\Author;
+use APP\facades\Repo;
+use Illuminate\Events\Dispatcher;
+use Illuminate\Support\Enumerable;
+use Illuminate\Support\Facades\Mail;
+use PKP\db\DAORegistry;
+use PKP\log\SubmissionEmailLogEntry;
+use PKP\mail\Mailable;
+use PKP\mail\mailables\SubmissionAcknowledgement;
+use PKP\mail\mailables\SubmissionAcknowledgementOtherAuthors;
+use PKP\observers\events\SubmissionSubmitted;
+use PKP\security\Role;
+use PKP\stageAssignment\StageAssignmentDAO;
+use PKP\user\User;
+
+abstract class SendSubmissionAcknowledgement
+{
+    abstract public function subscribe(Dispatcher $events): void;
+
+    public function handle(SubmissionSubmitted $event)
+    {
+        /** @var StageAssignmentDAO $stageAssignmentDao */
+        $stageAssignmentDao = DAORegistry::getDAO('StageAssignmentDAO');
+        $result = $stageAssignmentDao->getBySubmissionAndRoleIds($event->submission->getId(), [Role::ROLE_ID_AUTHOR]);
+        $assignedUserIds = [];
+        while ($stageAssignment = $result->next()) {
+            /** @var StageAssignment $stageAssignment */
+            $assignedUserIds[] = $stageAssignment->getUserId();
+        }
+
+        $submitterUsers = !empty($assignedUserIds)
+            ? Repo::user()
+                ->getCollector()
+                ->filterByUserIds($assignedUserIds)
+                ->getMany()
+            : collect();
```

Comment:
This is where writing the collectors to distinguish between `null` (filter not set) and `[]` (filter set, but with no acceptable values) will pay off. You should be able to just pass in `filterByUserIds($assignedUserIds)` even if it's an empty array and the filter will behave as expected. The ternary isn't necessary.

Reply:
:+1: This collector is coded that way so it will work. The one thing the ternary does is avoid an unnecessary DB query. But since it will return 0 results it's probably not a big performance gain. I've removed the ternary condition.